staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/typeconverter.go

```diff
@@ -36,30 +38,54 @@ type TypeConverter interface {
 }
 
 type typeConverter struct {
-	parser *managedfields.GvkParser
+	parser map[schema.GroupVersionKind]*typed.ParseableType
 }
 
 var _ TypeConverter = &typeConverter{}
 
-// NewTypeConverter builds a TypeConverter from a proto.Models. This
+// NewTypeConverter builds a TypeConverter from a spec.Swagger. This
 // will automatically find the proper version of the object, and the
 // corresponding schema information.
 func NewTypeConverter(openapiSpec *spec.Swagger, preserveUnknownFields bool) (TypeConverter, error) {
-	models, err := utilopenapi.ToProtoModels(openapiSpec)
-	if err != nil {
-		return nil, err
+	if openapiSpec == nil {
+		return nil, errors.New("openapi spec must not be nil")
+	}
+
+	pointerDefs := map[string]*spec.Schema{}
+	for k, v := range openapiSpec.Definitions {
+		vCopy := v
+		pointerDefs[k] = &vCopy
 	}
-	parser, err := managedfields.NewGVKParser(models, preserveUnknownFields)
+
+	typeSchema, err := schemaconv.ToSchemaFromOpenAPI(pointerDefs, preserveUnknownFields)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to convert models to schema: %v", err)
 	}
-	return &typeConverter{parser: parser}, nil
+
+	typeParser := typed.Parser{Schema: smdschema.Schema{Types: typeSchema.Types}}
+
+	tr := map[schema.GroupVersionKind]*typed.ParseableType{}
+	for modelName, model := range pointerDefs {
+		gvkList := parseGroupVersionKind(model.Extensions)
+		if len(gvkList) == 0 {
+			continue
+		}
+
+		parsedType := typeParser.Type(modelName)
+		for _, gvk := range gvkList {
+			if len(gvk.Kind) > 0 {
+				tr[schema.GroupVersionKind(gvk)] = &parsedType
+			}
+		}
+	}
```

Comment:
Not super strong opinion but you could make that in a separate function, not asking for a change though.